pkg/deploy/dev-workspace-config/dev_workspace_config.go

```diff
@@ -71,52 +71,63 @@ func (d *DevWorkspaceConfigReconciler) Finalize(ctx *chetypes.DeployContext) boo
-		}
+	pvcStrategy := utils.GetValue(devEnvironments.Storage.PvcStrategy, constants.DefaultPvcStorageStrategy)
+	isPerWorkspacePVCStorageStrategy := pvcStrategy == constants.PerWorkspacePVCStorageStrategy
+	pvcConfig := map[bool]*chev2.PVC{
+		true:  devEnvironments.Storage.PerWorkspaceStrategyPvcConfig,
+		false: devEnvironments.Storage.PerUserStrategyPvcConfig,
+	}[isPerWorkspacePVCStorageStrategy]
+
+	if err := updateWorkspaceStorageConfig(pvcConfig, isPerWorkspacePVCStorageStrategy, operatorConfig.Workspace); err != nil {
+		return err
 	}
 
-	if pvc != nil {
-		if operatorConfig.Workspace == nil {
-			operatorConfig.Workspace = &controllerv1alpha1.WorkspaceConfig{}
-		}
-		return updateWorkspaceConfig(pvc, pvcStrategy == constants.PerWorkspacePVCStorageStrategy, operatorConfig.Workspace)
+	if err := updateWorkspaceServiceAccountConfig(devEnvironments, operatorConfig.Workspace); err != nil {
+		return err
 	}
+
 	return nil
 }
 
-func updateWorkspaceConfig(pvc *chev2.PVC, isPerWorkspacePVCStorageStrategy bool, workspaceConfig *controllerv1alpha1.WorkspaceConfig) error {
-	if pvc.StorageClass != "" {
-		workspaceConfig.StorageClassName = &pvc.StorageClass
-	}
-
-	if pvc.ClaimSize != "" {
-		if workspaceConfig.DefaultStorageSize == nil {
-			workspaceConfig.DefaultStorageSize = &controllerv1alpha1.StorageSizes{}
+func updateWorkspaceStorageConfig(pvc *chev2.PVC, isPerWorkspacePVCStorageStrategy bool, workspaceConfig *controllerv1alpha1.WorkspaceConfig) error {
+	if pvc != nil {
+		if pvc.StorageClass != "" {
+			workspaceConfig.StorageClassName = &pvc.StorageClass
 		}
 
-		pvcSize, err := resource.ParseQuantity(pvc.ClaimSize)
-		if err != nil {
-			return err
+		if pvc.ClaimSize != "" {
+			if workspaceConfig.DefaultStorageSize == nil {
+				workspaceConfig.DefaultStorageSize = &controllerv1alpha1.StorageSizes{}
+			}
+
+			pvcSize, err := resource.ParseQuantity(pvc.ClaimSize)
+			if err != nil {
+				return err
+			}
+
+			if isPerWorkspacePVCStorageStrategy {
+				workspaceConfig.DefaultStorageSize.PerWorkspace = &pvcSize
+			} else {
+				workspaceConfig.DefaultStorageSize.Common = &pvcSize
+			}
 		}
+	}
+	return nil
+}
 
-		if isPerWorkspacePVCStorageStrategy {
-			workspaceConfig.DefaultStorageSize.PerWorkspace = &pvcSize
-		} else {
-			workspaceConfig.DefaultStorageSize.Common = &pvcSize
-		}
+func updateWorkspaceServiceAccountConfig(devEnvironments *chev2.CheClusterDevEnvironments, workspaceConfig *controllerv1alpha1.WorkspaceConfig) error {
+	isNamespaceAutoProvisioned := pointer.BoolPtrDerefOr(devEnvironments.DefaultNamespace.AutoProvision, constants.DefaultAutoProvision)
+
+	workspaceConfig.ServiceAccount = &controllerv1alpha1.ServiceAccountConfig{
+		ServiceAccountName: devEnvironments.ServiceAccount,
+		// If user's Namespace is not auto provisioned (is pre-created by admin),
+		// then ServiceAccount must be pre-created as well
+		DisableCreation: pointer.BoolPtr(!isNamespaceAutoProvisioned),
```

Comment:
From what I understood [1] here, if namespace is pre-created, then SA must be pre-created as well.
And the SA name must be defined in CheCluster CR to reuse by DWO.

[1] https://issues.redhat.com/browse/CRW-3470

Reply:
I think I confused myself in the initial comment, but there seems to be something strange about how this is configured (though I'm not sure what it is at the moment :thinking:)

The description in that Jira is somewhat unclear. The valid configurations on the DWO side are:

1. No config specified, per-workspace SAs are used. This is the default:
    ```yaml
    config:
      workspace:
        serviceAccount: {}
    ```

2. Custom SA name configured, automatically created by DWO when a DevWorkspace is created and shared by all workspaces in the namespace (this SA does not get automatically cleaned up):
    ```yaml
    config:
      workspace:
        serviceAccount:
          serviceAccountName: <SA_NAME>
    ```

3. Custom SA name configured and automatic creation disabled; it's assumed that someone else (e.g. an admin) will create the SA for each namespace:
    ```yaml
    config:
      workspace:
        serviceAccount:
          disableCreation: true
          serviceAccountName: <SA_NAME>
    ```

I think I was assuming our goals were to support cases 1 and 3, since case 2 is a bit of an edge case (you want one SA name for all workspaces, but don't want to make it yourself, but also don't mind cleaning up automatically-created resources).

Reply:
> Custom SA name configured and automatic creation disabled; it's assumed that someone else (e.g. an admin) will create the SA for each namespace:

yes, that's my understanding as well.